File: sc32v3_stm32/lib/common/SerialProtocol.cpp
#include "SerialProtocol.h"
#include "crc.h"
#define EMPTY_BYTE 0x55

SerialProtocol::SerialProtocol(Stream* stream)
{
    this->stream = stream;
    memset(frame_buf, EMPTY_BYTE, sizeof(frame_buf));
}

void SerialProtocol::write(uint16_t header, void* data, uint16_t size)
{
    uint16_t checksum = get_CRC16(data, size);
    stream->write((uint8_t*)&header, sizeof(uint16_t));
    stream->write((uint8_t*)&size, sizeof(uint16_t));
    stream->write((uint8_t*)data, size);
    stream->write((uint8_t*)&checksum, sizeof(uint16_t));
}

bool SerialProtocol::read(uint16_t header, void* data, uint16_t size) {
    if (header == 0 || data == nullptr || size > (SERIAL_PROTOCOL_MAX_BUFFER_SIZE >> 1)) return false;
    frame_buf_read(size);
    return parse_frame(header, data, size);
}

bool SerialProtocol::parse_frame(uint16_t header, void* result_data, uint16_t expected_data_size)
{
    if (frame_buf_len > 0)
    {
        uint16_t rx_buf_max_len = (expected_data_size+6);
        uint16_t frame_buf_max_len = rx_buf_max_len << 1;
        int frame_start_offset = 0;
        bool header_found = false;
        while (frame_start_offset < (frame_buf_len - 4)) // offset should include at least 4 bytes of [header]+[data len]
        {
            uint16_t* hdr = (uint16_t*)&frame_buf[frame_start_offset];
            if (*hdr == header)
            {
                header_found = true;
                break;
            }
            frame_start_offset++;
        }
        if (header_found)
        {
            // align frame start with 0 start index in buffer
            move_frame(frame_start_offset, frame_buf_max_len);

            //process frame if buffer len is as expected + 6 bytes = [header](2)+[data len](2)+[crc](2)
            if (frame_buf_len >= rx_buf_max_len)
            {
                uint16_t data_size = *(uint16_t*)&frame_buf[2]; //skip 2 bytes of [header]
                if (data_size == expected_data_size)
                {
                    uint16_t expected_crc = get_CRC16(&frame_buf[4], data_size);  //skip 2 bytes of [data len]
                    uint16_t crc = *(uint16_t*)&frame_buf[4 + data_size];
                    if (crc == expected_crc)
                    {
                        memcpy(result_data, &frame_buf[4], data_size);
                        frame_buf_len = 0;
                        return true;
                    }
                }
            }
        }
    }
    return false;
}

uint16_t SerialProtocol::stream_read(uint8_t* buffer, uint16_t size)
{
    size_t avail = stream->available();
    if (size < avail) {
        avail = size;
    }
    size_t count = 0;
    while (count < avail) {
        *buffer++ = stream->read();
        count++;
    }
    stream->flush();
    return count;
}

uint16_t SerialProtocol::frame_buf_read(uint16_t expected_data_size)
{
    uint16_t rx_buf_max_len = (expected_data_size+6);
    uint16_t frame_buf_max_len = rx_buf_max_len << 1;
    size_t rx_buf_len = stream_read(rx_buf, rx_buf_max_len);
    if (rx_buf_len <= 0)
        return rx_buf_len;
    else
    {
        if ((frame_buf_len + rx_buf_len) > frame_buf_max_len)
        {
            // data overflow
            frame_buf_len = 0;
        }
        memcpy(frame_buf + frame_buf_len, rx_buf, rx_buf_len);
        frame_buf_len += rx_buf_len;
        return rx_buf_len;
    }
    return 0;
}

void SerialProtocol::move_frame(uint16_t offset, uint16_t size)
{
    if (offset <= 0) return;
    memmove(frame_buf, &frame_buf[offset], frame_buf_len - offset);
    frame_buf_len = frame_buf_len - offset;
    if ((size - frame_buf_len) >= 1)
        memset(frame_buf + frame_buf_len, EMPTY_BYTE, size - frame_buf_len);
}

File: HexaLink/PcLink/Libs/SerialProtocol.cs
using SerialPortLib2;
using System;
using System.Diagnostics;
using System.IO;
using System.Threading;
using System.Threading.Tasks;

public class SerialProtocol: IDisposable
{
    private SerialPortInput serialPort = new SerialPortInput(false);
    private byte[] frame_buf = new byte[1024*4];
    private int frame_buf_len = 0;

    public delegate void FrameReadyEventHandler(object sender, FrameReadyEventArgs e);
    public event FrameReadyEventHandler OnFrameReady;

    Stopwatch sw = new Stopwatch();
    Stopwatch fpssw = new Stopwatch();
    int fpsc = 0, fps = 0;

    public int FPS => fps;

    CancellationTokenSource tokenSource;
    CancellationToken token;
    Task task;

    public void Start()
    {
        sw.Start();
        fpssw.Start();
        serialPort.SetPort("COM3", 115200, readerTaskTime: 25);
        serialPort.ConnectionStatusChanged += SerialPort_ConnectionStatusChanged;
        serialPort.MessageReceived += SerialPort_MessageReceived;
        serialPort.Connect();
        
        tokenSource = new CancellationTokenSource();
        token = tokenSource.Token;
        task = Task.Run(() =>
        {
            while (true)
            {
                Loop();
                if (token.IsCancellationRequested)
                {
                    break;
                }
            }
        });
    }

    public void Stop()
    {
        serialPort.Disconnect();

        tokenSource.Cancel();
        task.Wait();
        task.Dispose();
    }

    public void SendFrame(FrameHeaderType header, byte[] data)
    {
        MemoryStream ms = new MemoryStream();
        BinaryWriter bw = new BinaryWriter(ms);

        bw.Write((ushort)header);
        bw.Write((ushort)data.Length);
        bw.Write(data);
        bw.Write(Crc.Get_CRC16(data));

        serialPort.SendMessage(ms.ToArray());
    }

    public void Dispose()
    {
        Stop();
    }

    private void Update(byte[] rx_buf)
    {
        int rx_len = rx_buf.Length;
        if (rx_len > 0)
        {
            if ((frame_buf_len + rx_len) > frame_buf.Length)
            {
                // data overflow
                frame_buf_len = 0;
            }
            Buffer.BlockCopy(rx_buf, 0, frame_buf, frame_buf_len, rx_len);

            frame_buf_len += rx_len;
        }
    }

    private void Loop()
    {
        if (sw.ElapsedMilliseconds < 10) return;
        if (frame_buf_len > 0)
        {
            var frame_br = new BinaryReader(new MemoryStream(frame_buf));
            int frame_start_offset = 0;
            FrameHeaderType header = FrameHeaderType.Unknown;
            while (frame_start_offset < (frame_buf_len - 4))
            {
                frame_br.BaseStream.Seek(frame_start_offset, SeekOrigin.Begin);
                header = (FrameHeaderType)frame_br.ReadUInt16();
                if (header == FrameHeaderType.STM32Debug ||
                    header == FrameHeaderType.ESP32Debug)
                {
                    break;
                }
                else
                {
                    header = FrameHeaderType.Unknown;
                }
                frame_start_offset++;
            }
            if (header != FrameHeaderType.Unknown)
            {
                // align frame start with 0 start index in buffer
                MoveFrame(frame_start_offset, (ushort)frame_buf.Length);
                ushort expectedDataSize = GetExpectedDataSize(header);
                if (frame_buf_len >= expectedDataSize)
                {
                    frame_br.BaseStream.Seek(2, SeekOrigin.Begin);
                    ushort data_size = frame_br.ReadUInt16();
                    if (data_size == expectedDataSize)
                    {
                        var data = frame_br.ReadBytes(data_size);
                        var expected_crc = Crc.Get_CRC16(data);
                        var actual_crc = frame_br.ReadUInt16();
                        bool crc_valid = actual_crc == expected_crc;
                        if (crc_valid)
                        {
                            OnFrameReady(this, FrameReadyEventArgsBuilder.Create(header, data));
                            frame_buf_len = 0;
                        }
                    }
                }
            }
        }
        sw.Restart();
    }

    private void MoveFrame(int offset, ushort size)
    {
        if (offset <= 0) return;
        Buffer.BlockCopy(frame_buf, offset, frame_buf, 0, frame_buf_len - offset);
        frame_buf_len = frame_buf_len - offset;
        if ((size - frame_buf_len) >= 1)
            for (int i = 0; i < (size - frame_buf_len); i++)
                frame_buf[i + frame_buf_len] = 0x55;
    }

    private ushort GetExpectedDataSize(FrameHeaderType header)
    {
        switch (header)
        {
            case FrameHeaderType.ESP32Debug:
                return 26 * 4 + 8 * 3 * 3 + 1 + 4;
            case FrameHeaderType.STM32Debug:
                return 26 * 4;
        }
        return 0;
    }

    private void SerialPort_MessageReceived(object sender, MessageReceivedEventArgs args)
    {
        if (fpssw.ElapsedMilliseconds >= 1000)
        {
            fps = fpsc;
            fpsc = 0;
            fpssw.Restart();
        }
        fpsc++;
        Update(args.Data);
    }

    private void SerialPort_ConnectionStatusChanged(object sender, ConnectionStatusChangedEventArgs args)
    {
    }
}